reference-implementation/src/header.rs

```diff
@@ -0,0 +1,149 @@
+/// An SFrame key ID
+#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
+pub struct KeyId(pub u64);
+
+/// An SFrame counter
+#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
+pub struct Counter(pub u64);
+
+/// The SFrame header.  It stores KID and CTR values for a ciphertext, and the encoded form.
+#[derive(PartialEq, Eq, Debug)]
+pub struct Header {
+    /// The key ID for the key with which the ciphertext was encrypted
+    pub kid: KeyId,
+
+    /// The counter used to form a unique nonce for the ciphertext
+    pub ctr: Counter,
+
+    encoded: Vec<u8>,
+}
+
+fn to_min_be_bytes(val: u64) -> Vec<u8> {
+    let bytes = val.to_be_bytes();
+    let start = bytes.iter().position(|&x| x != 0).unwrap_or(7);
+    bytes[start..].to_vec()
+}
+
+fn parse_be_bytes(data: &[u8]) -> u64 {
+    let mut bytes = [0u8; 8];
+    let start = bytes.len() - data.len();
+    bytes[start..].copy_from_slice(data);
+    u64::from_be_bytes(bytes)
+}
+
+impl Header {
+    /// Encode a new SFrame header with the specified KID and CTR values
+    pub fn new(kid: KeyId, ctr: Counter) -> Self {
+        let mut encoded = vec![0];
+
+        if kid.0 < 0x08 {
+            encoded[0] |= kid.0 as u8;
```

Comment:
```suggestion
            encoded[0] = kid.0 as u8;
```
though I would use `Vec::new()` above and `.push(kid.0 as u8)` here instead.

Reply:
It seems like allocating a one-element vector reflects the fact that there's always a config byte at the front of the header.  Maybe this could be more explicit as something like:

``` rust
let mut encoded = vec![0];
let config_byte = &mut encoded[0];
// ...
*config_byte = kid.0 as u8
```